### PR TITLE
Nfr adding authentication

### PR DESCRIPTION
**Pull request by eoinmcbennett**
Implemented authentication

### CODE REVIEW

**Comment by eoinmcbennett** [comment]
@steveh2707 could you give this quick review?

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean registerUser(Credential credential) throws GenericActionFailedException, AuthenticationException {
+        try {
+            Connection conn = DatabaseConnector.getConnection();
+            System.out.println(credential.getUsername() + " " + credential.getPassword());
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: remove println

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+            st.setString(1,credential.getUsername());
+
+            ResultSet rs = st.executeQuery();
+
+            while(rs.next()) {
+                return rs.getString("Password").equals(credential.getPassword());
+            }
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: throw error for consistency

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+        return false;
+    }
+
+    @Override
+    public boolean registerUser(Credential credential) throws GenericActionFailedException, AuthenticationException {
+        try {
+            Connection conn = DatabaseConnector.getConnection();
+            System.out.println(credential.getUsername() + " " + credential.getPassword());
+            String SQL = "INSERT INTO User(Username,Password,RoleID) VALUES(?,?,2)";
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: remove hardcoded 2

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            PreparedStatement st = conn.prepareStatement(SQL, Statement.RETURN_GENERATED_KEYS);
+
+            st.setString(1,credential.getUsername());
+            st.setString(2,credential.getPassword());
+
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/db/AuthDao.java:45]: add setInt for RoleID

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            PreparedStatement st = conn.prepareStatement(SQL);
+
+            st.setString(1,token);
+            ResultSet rs = st.executeQuery();
+
+            while(rs.next()) {
+                Timestamp expiry = rs.getTimestamp("Expiry");
+
+                if(expiry.after(new Date())) {
+                    return UserRole.values()[rs.getInt(1) - 1];
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: can you split this over 2 lines with comments to improve clarity

File: src/main/java/org/kainos/ea/resources/EmployeeController.java [hunk 6]
```diff
@@ -1,25 +1,41 @@
...
 @Path("/api")
 public class EmployeeController {
     private final EmployeeService employeeService = new EmployeeService();
+    private final AuthService authService = new AuthService(new AuthDao());
 
     @GET
     @Path("/employees")
     @Produces(MediaType.APPLICATION_JSON)
-    public Response getAllOrders() {
+    public Response getAllOrders(@QueryParam("token") String token) {
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/resources/EmployeeController.java:28]: change to getAllEmployees

File: src/main/java/org/kainos/ea/resources/EmployeeController.java [hunk 7]
```diff
@@ -54,7 +81,18 @@ public Response createOrder(DeliveryEmployeeRequest deliveryEmployee) {
...
+        try{
+            if(!authService.doesTokenHaveRole(token,UserRole.HR)){
+                return Response.status(Response.Status.FORBIDDEN).build();
+            }
+        } catch (GenericActionFailedException e) {
+            return Response.serverError().build();
+        } catch (AuthenticationException e) {
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
+        }
+
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/resources/EmployeeController.java:82]: combine these into 1 try block

File: src/main/java/org/kainos/ea/resources/EmployeeController.java [hunk 8]
```diff
@@ -54,7 +81,18 @@ public Response createOrder(DeliveryEmployeeRequest deliveryEmployee) {
...
     @Produces(MediaType.APPLICATION_JSON)
-    public Response getDeliveryEmployeeById(@PathParam("id") int id) {
+    public Response getDeliveryEmployeeById(@PathParam("id") int id,@QueryParam("token") String token) {
+
+        try{
+            if(!authService.doesTokenHaveRole(token,UserRole.HR)){
+                return Response.status(Response.Status.FORBIDDEN).build();
+            }
+        } catch (GenericActionFailedException e) {
+            return Response.serverError().build();
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/resources/EmployeeController.java]: log exceptions before throwing

File: src/main/java/org/kainos/ea/resources/EmployeeController.java [hunk 9]
```diff
@@ -73,7 +111,18 @@ public Response getDeliveryEmployeeById(@PathParam("id") int id) {
...
+        } catch (AuthenticationException e) {
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
+        }
+
         try {
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/resources/EmployeeController.java]: as before

File: src/main/java/org/kainos/ea/resources/EmployeeController.java [hunk 10]
```diff
@@ -88,21 +137,50 @@ public Response getAllDeliveryEmployees(){
...
+    public Response deleteDeliveryEmployee(@PathParam("id") int id, @QueryParam("token") String token) {
+        try{
+            if(!authService.doesTokenHaveRole(token,UserRole.HR)){
+                return Response.status(Response.Status.FORBIDDEN).build();
+            }
+        } catch (GenericActionFailedException e) {
+            return Response.serverError().build();
+        } catch (AuthenticationException e) {
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
+        }
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/resources/EmployeeController.java]: as before

File: src/main/java/org/kainos/ea/resources/ProjectController.java [hunk 11]
```diff
@@ -1,31 +1,45 @@
...
+                return Response.status(Response.Status.FORBIDDEN).build();
+            }
+        } catch (GenericActionFailedException e) {
+            System.err.println(e.getMessage());
+            return Response.serverError().build();
+        } catch (AuthenticationException e) {
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
+        }
+
         try {
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/resources/ProjectController.java]: as before

File: src/main/java/org/kainos/ea/resources/ProjectController.java [hunk 12]
```diff
@@ -45,7 +58,18 @@ public Response addDeliveryEmployeeToProject(@PathParam("id") int id, ProjectAdd
...
+        try{
+            if(!authService.doesTokenHaveRole(token,UserRole.MANAGEMENT)){
+                return Response.status(Response.Status.FORBIDDEN).build();
+            }
+        } catch (GenericActionFailedException e) {
+            return Response.serverError().build();
+        } catch (AuthenticationException e) {
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
+        }
         try {
```
Comment (steveh2707) [inline src/main/java/org/kainos/ea/resources/ProjectController.java]: as before

File: src/main/java/org/kainos/ea/db/AuthDao.java (hunk 3 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/db/AuthDao.java]: We actually need this to default a user to a specific role. We might want to consider a change in the SQL to default this value so it doesn't need to be specified?

File: src/main/java/org/kainos/ea/db/AuthDao.java (hunk 4 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/db/AuthDao.java:45]: https://github.com/steveh2707/soSOLIDcrew/pull/2#discussion_r1372999203

File: src/main/java/org/kainos/ea/db/AuthDao.java (hunk 5 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/db/AuthDao.java]: Yeah two lines is a lot clearer. Added a comment here just to be sure

File: src/main/java/org/kainos/ea/db/AuthDao.java (hunk 3 above)
Comment (steveh2707) [reply src/main/java/org/kainos/ea/db/AuthDao.java]: maybe just add the hardcoded value in the setInt function to make it easy to update later.

Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/db/AuthDao.java]: Yeah good shout, sorted

File: src/main/java/org/kainos/ea/db/EmployeeDao.java [hunk 13]
```diff
@@ -133,6 +131,24 @@ public int createDeliveryEmployee(DeliveryEmployeeRequest deliveryEmployee) thro
...
+
+        String updateStatement = "UPDATE employee SET first_name = ?, last_name = ?, salary = ?, bank_account_number = ?, national_insurance_number = ? WHERE employee_id = ?";
+
+        PreparedStatement st = c.prepareStatement(updateStatement);
+
+        st.setString(1, employee.getFirstName());
+        st.setString(2, employee.getLastName());
+        st.setDouble(3, employee.getSalary());
+        st.setString(4,employee.getBankAccountNumber());
+        st.setString(5,employee.getNationalInsuranceNumber());
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/db/EmployeeDao.java]: shouldn't be able to update NIN for delivery employee (create a new model)

File: src/main/java/org/kainos/ea/resources/AuthController.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+    @POST
+    @Path("/login")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response login(Credential login){
+        try {
+            String token = authService.login(login);
+            if(token == null){
+                return Response.status(Response.Status.BAD_REQUEST).build();
+            }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/AuthController.java:30]: move this logic to the service, throw an exception that can be caught below. also, I would have assumed the token being null was a problem on our end so should return 500?

**Review by shaunganley — APPROVED** [review]
approved assuming comment will be addressed in previous PR

**Pull request merged**